### PR TITLE
Multichannel fixes for reconnect scenarios

### PR DESCRIPTION
**Pull request by sprasad-microsoft**
Made a bunch of fixes in mchan reconnect scenarios with these patches. The reconnect codepath did not play well with mchan. Making it work involved several changes described in the commits.

https://github.com/sprasad-microsoft/smb-kernel-client/commit/03059a751fa7315ecd44ad61342200d20ab3f0cb.patch
A trivial ref-counting fix in session find codepath. 

https://github.com/sprasad-microsoft/smb-kernel-client/commit/2615c7300f93a7ae2d9350da25ed117492f8edbf.patch
Fixes in reconnect codepath, involving changes to introduce a per-channel bitmask for a session, instead of a boolean, and to reconnect the session only when we have all channels marked for reconnect. This makes sessions more resilient in mchan scenarios.

https://github.com/sprasad-microsoft/smb-kernel-client/commit/5bcce1741beea482e770f1c25f4ff285a1505ca4.patch
Fix to get rid of the serialization of requests during channel binding. This involved passing the channel server pointer to all negotiate and session setup related functions. This has increased the diff size, otherwise the changes are quite minimal.

### CODE REVIEW

File: fs/cifs/cifssmb.c [hunk 1]
```diff
@@ -217,8 +221,19 @@ cifs_reconnect_tcon(struct cifs_tcon *tcon, int smb_command)
 		goto out;
 	}
 
+	/* Recheck after acquire mutex */
+	if (!cifs_chan_needs_reconnect(ses, server)) {
+		/* this just means that we only need to tcon */
```
Comment (aaptel) [inline fs/cifs/cifssmb.c:226]: missing a word

File: fs/cifs/sess.c [hunk 2]
```diff
@@ -256,15 +300,16 @@ cifs_ses_add_channel(struct cifs_sb_info *cifs_sb, struct cifs_ses *ses,
 	mutex_lock(&ses->session_mutex);
 
 	chan = ses->binding_chan = &ses->chans[ses->chan_count];
-	chan->server = cifs_get_tcp_session(&ctx);
+	/* Indiacate ses->server as the primary server */
+	chan->server = cifs_get_tcp_session(&ctx, ses->server);
```
Comment (aaptel) [inline fs/cifs/sess.c:304]: typo

**Review by aaptel — COMMENTED** [review]
We have already discussed the general way to code this so I have no comment regarding the design. LGTM

Few style issues, see inline comments. Also, use about 100 columns as your max line length, you can remote some line breaks.

File: fs/cifs/sess.c [hunk 3]
```diff
@@ -1727,11 +1719,13 @@ sess_auth_rawntlmssp_authenticate(struct sess_data *sess_data)
 	sess_data->result = rc;
 }
 
-static int select_sec(struct cifs_ses *ses, struct sess_data *sess_data)
+static int select_sec(struct cifs_ses *ses,
+		      struct TCP_Server_Info *server,
+		      struct sess_data *sess_data)
 {
```
Comment (aaptel) [inline fs/cifs/sess.c:1725]: can we use sess_data->server here? I feel like we don't need the extra arg

File: fs/cifs/transport.c [hunk 4]
```diff
@@ -1198,7 +1199,8 @@ compound_send_recv(const unsigned int xid, struct cifs_ses *ses,
 	 */
 	if ((ses->status == CifsNew) || (optype & CIFS_NEG_OP) || (optype & CIFS_SESS_OP)) {
 		mutex_lock(&server->srv_mutex);
-		smb311_update_preauth_hash(ses, rqst[0].rq_iov,
+		smb311_update_preauth_hash(ses, server,
+					   rqst[0].rq_iov,
 					   rqst[0].rq_nvec);
```
Comment (aaptel) [inline fs/cifs/transport.c:1204]: no need for linebreak

File: fs/cifs/transport.c [hunk 5]
```diff
@@ -1270,7 +1272,8 @@ compound_send_recv(const unsigned int xid, struct cifs_ses *ses,
 			.iov_len = resp_iov[0].iov_len
 		};
 		mutex_lock(&server->srv_mutex);
-		smb311_update_preauth_hash(ses, &iov, 1);
+		smb311_update_preauth_hash(ses, server,
+					   &iov, 1);
 		mutex_unlock(&server->srv_mutex);
```
Comment (aaptel) [inline fs/cifs/transport.c:1277]: no need for linebreak

**Review by aaptel — COMMENTED** [review]
As we already discussed the design, this change LGTM
Minor remarks: see inline comments

As a note for future change, I suspect signkey logic can be simplified also. We can overwrite chan->server->signkey instead of having separate chan->signkey

File: fs/cifs/cifssmb.c (hunk 1 above)
Comment (sprasad-microsoft) [reply fs/cifs/cifssmb.c:226]: I meant that...
"this means that we only need to tree connect"
I'll fix it.

File: fs/cifs/sess.c (hunk 2 above)
Comment (sprasad-microsoft) [reply fs/cifs/sess.c:304]: Will fix.

File: fs/cifs/sess.c (hunk 3 above)
Comment (sprasad-microsoft) [reply fs/cifs/sess.c:1725]: Good catch. I remember fixing this in many other places. Must have missed here.
Will fix. Thanks.

File: fs/cifs/transport.c (hunk 4 above)
Comment (sprasad-microsoft) [reply fs/cifs/transport.c:1204]: Will fix.

File: fs/cifs/transport.c (hunk 5 above)
Comment (sprasad-microsoft) [reply fs/cifs/transport.c:1277]: Will fix.

**Comment by sprasad-microsoft** [comment]
Have raised a new PR with updated patches. Aborting this one.

**Pull request closed without merging**